Modules/_decimal/_decimal.c

```diff
@@ -3672,34 +3673,45 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
+
+    int64_t val = mpd_qget_i64(x, &status);
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+
+    const uint8_t bpd = layout->bits_per_digit;
+    n = (mpd_sizeinbase(x, 2) + bpd - 1) / bpd;
+    PyLongWriter *writer = PyLongWriter_Create(mpd_isnegative(x), n,
+                                               (void**)&ob_digit);
+    /* mpd_sizeinbase can overestimate size by 1 digit, set it to zero. */
```

Comment:
BTW, this looks as a bug in the mpdecimal.  C.f. the GNU GMP, the mpz_sizeinbase docs says: "If base is a power of 2, the result is always exact".